src/action-sheets/index.js

```diff
@@ -199,6 +206,48 @@ const copyLinkToTopic = {
   },
 };
 
+const toggleResolveTopic = async ({ auth, streamId, topic, _, streams, zulipFeatureLevel }) => {
+  // TODO: It'd be nice to use a message ID we know is in the conversation,
+  //   where possible, rather than do this extra fetch.  Just need to thread
+  //   such a message ID through to here.  (Like web, we'll still need this
+  //   for cases where we can show a topic without knowing of any message ID
+  //   definitely in it, e.g. in the list of topics for a stream.)
+  const messageId = await fetchSomeMessageIdForConversation(auth, streamId, topic, streams);
+  if (messageId == null) {
+    // The conversation is actually empty.  This can be perfectly normal,
+    // because by fetching information outside the events system we're
+    // exposed to a race: e.g., someone else could have resolved the topic
+    // already and we just hadn't heard yet.
+    showToast(_('No messages in topic'));
```

Comment:
Yeah. Someone could also have resolved/unresolved the topic during the time between the action sheet opening and the resolve/unresolve button being pressed.

I wonder if "No messages in topic" is quite the right message. The average user probably doesn't think of the with-checkmark and without-checkmark topics as two different topics, right?

Of course, it's possible that we got here because all the messages with the topic were deleted, or the topic's words changed. I think these are cases where "No messages in topic" might make some sense to the user. (Clearer, though, to say something more specific to those cases, I think.)

But if it's just that a resolve/unresolve has already been done, I wonder if "No messages in topic" will seem like a bug.

In this case, actually, might a really generic failure message be better? The action sheet will close on the failure (I see this experimentally). The event system will make sure the user is looking at up-to-date state. If the user isn't happy with that state and they want to do a resolve/unresolve action again, they'll do it by opening the action sheet again. That action-sheet session will act on a new snapshot of the state, and this same failure is pretty unlikely to happen on the retry.

Reply:
I wouldn't worry too much about this—better to have the feature in the first place than not. 🙂

Whatever message we settle on for the toast, though, it should get an entry in messages_en.json.

Reply:
Hmm, yeah. I did some manual testing while inducing artificial failures here, and I found what I think is a good solution -- take a look.



> Whatever message we settle on for the toast, though, it should get an entry in messages_en.json.

Indeed -- thanks for the catch!